### DISCUSSION TITLE
2.0 constraints

### DISCUSSION THREAD

**Discussion by nlemoine** [timber/timber#2436] [General]
Hi Timber team, 

I'd like to open a discussion on PHP/WP/Twig minimum requirements here because this is a topic that is mentioned is various issues/PR but isn't centralized somewhere.

Related: #2407 / https://github.com/timber/timber/pull/2349#issuecomment-760306256 / #1863 / #2339

### Twig

Although `twig:^2.10` is the current Twig constraint in Timber, it actually [doesn't pass tests](https://github.com/nlemoine/timber/actions/runs/738194552) (this is `2.x` tested with GA).

From @jarednova : 

> we're currently at ^2.10 in the 2.x branch. I'm all for going to ^2.10 || 3.0 as a 2.x Future thing.

Problem: `twig/cache-extension` has a `twig:^2.4` constraint and it would be a shame to release the brand new Timber 2.0 with an extension that's already not maintained anymore.

### PHP

I feel like the situation has really changed over the one/two last years. Requiring PHP `^7.4` is quite common these days.  Package's developers don't bother supporting a wide range of PHP version anymore. 

Even WordPress is now pushing forward latest PHP version adoption ^^ And every decent hosting company makes the latest PHP versions available almost at the same time they are released. 

Just like Twig, Timber 2.x actually [doesn't pass tests](https://github.com/nlemoine/timber/actions/runs/738194552) on PHP 7.0 or 7.1. Requiring an up to date version would also simplify testing. 

### Upgrade policy 

On a more global upgrade policy point of view, you are being very careful about backwards compatibility/not introducing BCs. This is nice for users who don't follow Timber feed like I do 😜 

But I think it's holding a lot of things back. Timber has a lot of legacy code (I didn't realize it was that much before recently digging into it) and I feel like 2.0 is a great opportunity to introduce breaking changes/catch up with latest practices/requirements. An opportunity that doesn't happen that often considering Timber lifecycle (1.0 was released almost exactly 5 years ago!). I'm afraid too much `2.x future` labels will postpone important things too far in the future (2.x is already in the pipes for quite a time now).

Don't get me wrong, this isn't in any way some hard criticism on your work/involvement/timeline on Timber's development. I'm fully aware of the hard work/time you put into it and the burden that maintaining OSS sometimes is. I also might not have all cards in hand to accurately evaluate your upgrade policy/decisions regarding breaking changes introduction, etc.

This just my humble opinion on this, opened for discussion obviously 😉 

Wherever it leads, I think there's still an objective problem between stated support and actual support for those ones (I'll be working again on GA tests soon).

**Comment by gchtr** [comment]
Sorry for the very late reply. And thanks for opening this discussion. It’s good to move this out of issues and pull request comments to a more central place.

I totally agree with you with everything you mentioned. And I still feel what I said in https://github.com/timber/timber/issues/1863#issuecomment-760493487 applies.

> @acobster Quite some passed has passed again. I think a good reference is always the [versions](https://www.php.net/supported-versions.php) page, where we see that active support for PHP 7.3 has already ended. The [WordPress requirements](https://wordpress.org/about/requirements/) lists PHP 7.4 as the recommended version as well. And as you said yourself in [#2339 (comment)](https://github.com/timber/timber/issues/2339#issuecomment-709600634), 7.4 isn’t such a big step from 7.3. So, I would **feel good about going with 7.4 for Timber 2.x**.

As mentioned in https://github.com/timber/timber/issues/2407, I’m not sure we even need to switch to `twig/cache-extra`. But we should still switch to `2.4` if we still need to support `twig/cache-extension`.

When I quickly looked at the [changelog for twigphp/twig](https://github.com/twigphp/Twig/blob/3.x/CHANGELOG), I didn’t see anything alarming when we upgrade to a minimum of `2.4`.

> Just like Twig, Timber 2.x actually [doesn't pass tests](https://github.com/nlemoine/timber/actions/runs/738194552) on PHP 7.0 or 7.1. Requiring an up to date version would also simplify testing.

I feel like tests only hold us back for now, also because it would be cool if we could switch to GitHub Actions and make the automated tests work again. You already put a [lot of work into this](https://github.com/timber/timber/pull/2417) and I would really like to see this fly.

So the following steps would be required:

- Update minimum required PHP version in composer.json to 7.4:
	`"php": "^7.4"`
- Update minimum Twig required version in composer.json to `^2.4|^3.0"`:
	`"twig/twig": "^2.4|^3.0"`

How do you feel about opening a pull request for this?

> But I think it's holding a lot of things back. Timber has a lot of legacy code (I didn't realize it was that much before recently digging into it) and I feel like 2.0 is a great opportunity to introduce breaking changes/catch up with latest practices/requirements. An opportunity that doesn't happen that often considering Timber lifecycle (1.0 was released almost exactly 5 years ago!). I'm afraid too much `2.x future` labels will postpone important things too far in the future (2.x is already in the pipes for quite a time now).

Heck yes, we also feel like it’s taking too long, but there are soo many things to consider, as you realized yourself. But I also feel like we should work toward releasing a stable 2.0 version. And then we should release newer versions with breaking changes more confidently. I know that @jared also feels like this. We don’t have to gather a lot of breaking changes for a version 3, but could also release them in versions 4, 5, etc. The `2.x future` tag only means that we don’t have to do it for `2.0`, be we also don’t know in which Timber version we’re going to solve it.

> Don't get me wrong, this isn't in any way some hard criticism on your work/involvement/timeline on Timber's development. I'm fully aware of the hard work/time you put into it and the burden that maintaining OSS sometimes is. I also might not have all cards in hand to accurately evaluate your upgrade policy/decisions regarding breaking changes introduction, etc.

It doesn’t come over as criticism at all :). These are all valid concerns and you’re allowed to mention them. I myself would love to find more time, because I’m still very motivated to work on Timber. So thanks for understanding if some of us still put private life and not Open Source first.

**Comment by nlemoine** [comment]
Hi @gchtr,

Thanks for your reply! 

> I’m not sure we even need to switch to twig/cache-extra

That would be a shame for many reasons. 
- As I previously said, releasing Timber 2 with an already abandoned (and unmaintained ?) package is not ideal but from a strict POV, yes, twig/cache-extension is still working fine and it could keep working for some time
- twig/cache-extra is a far better abstracted cache extension that lets you use any cache provider behind the scene (Redis, file system, etc.), including PSR 6/12 compliant packages, [some using WP transients already exist ](https://packagist.org/?query=psr%2016%20wordpress). 
- Another annoying thing is that [Timber does not let you use](https://github.com/timber/timber/blob/2.x/lib/Loader.php#L430) another cache extension/provider which could be useful if you're building some website with high cache expectations

> we also feel like it’s taking too long, but there are soo many things to consider, as you realized yourself

I did indeed but it's also a backwards compatibility policy choice (you could _choose_ to not have a backwards compatibility policy, at least not from 1.x). Timber has a lot of old code, some parts are still there [with no one really knowing why](https://github.com/timber/timber/pull/2397#issuecomment-765484348) 😅 . At the time it was designed, every method was public, there was a lot of public static properties, etc. IMO, deprecating each part of that old code is cumbersome and is, every all other legitimate reasons apart (private life, etc.), one of the reason things are moving slowly (it asks more work on code _and_ docs).

Considering no major update has been released since a long time and more importantly, now that Timber is a 100% PHP package (no more "click to update plugin" risk / people using composer should be aware that a major version can introduce breaking changes), I would be more in a "many things changed, read the docs and deal with it" mood 😄

I know you obviously don't share that point of view, you wouldn't work that hard to keep this v1 -> v2 transition that smooth 😉 So I'll deal with it and try to contribute when I can.
